File: nodejs/expressmvc/routes/merge-server-files.js
const express = require('express');
const { SampleDocs, StorageMock } = require('../storage-mock');

const router = express.Router();

class MergeServerFileModel {
	constructor(optionId, id1, id2, description) {
		this.optionId = optionId;
		this.id1 = id1;
		this.id2 = id2;
		this.description = description;
	}
}

router.get('/', (req, res) => {
	const availableFiles = [
		new MergeServerFileModel(0, SampleDocs.CMS_DETACHED_1, SampleDocs.CMS_DETACHED_2, 'A sample where both CMS don\'t have encapsulated content.'),
		new MergeServerFileModel(1, SampleDocs.CMS_ATTACHED_1, SampleDocs.CMS_DETACHED_2, 'A sample where a CMS has encapsulated content and another doesn\'t.'),
		new MergeServerFileModel(2, SampleDocs.CMS_ATTACHED_1, SampleDocs.CMS_ATTACHED_2, 'A sample where both CMS have encapsulated content.')];

	const sampleFileId = SampleDocs.CMS_DATA_FILE;

	res.render('merge-server-files', {
		availableFiles,
		sampleFileId
	});
});

router.post('/', (req, res) => {
	const sampleId = Number(req.body.selectedSample);

	let file1 = null;
	let file2 = null;
	const dataFile = StorageMock.readSampleDocSync(SampleDocs.CMS_DATA_FILE.valueOf());
	
	// Redirect the user to the signature route, passing the name of the file as
	// a URL argument.
	let redirectUrl = req.query.rc;
	
	switch (sampleId) {
	case 0:
		redirectUrl += `/detached?dataFileId=${StorageMock.storeSync(dataFile, null, '.pdf')}&`;
		file1 = StorageMock.readSampleDocSync(SampleDocs.CMS_DETACHED_1.valueOf());
		file2 = StorageMock.readSampleDocSync(SampleDocs.CMS_DETACHED_2.valueOf());
		break;
	case 1:
		redirectUrl += `/mixed?dataFileId=${StorageMock.storeSync(dataFile, null, '.pdf')}&`;
		file1 = StorageMock.readSampleDocSync(SampleDocs.CMS_ATTACHED_1.valueOf());
		file2 = StorageMock.readSampleDocSync(SampleDocs.CMS_DETACHED_2.valueOf());
		break;
	case 2:
		redirectUrl += '/attached?';
		file1 = StorageMock.readSampleDocSync(SampleDocs.CMS_ATTACHED_1.valueOf());
		file2 = StorageMock.readSampleDocSync(SampleDocs.CMS_ATTACHED_2.valueOf());
		break;
	default:
		break;
	}

	// Copy file to the App_Data folder, where the upload files is stored.
	const fileId1 = StorageMock.storeSync(file1, null, '.ps7');
	const fileId2 = StorageMock.storeSync(file2, null, '.ps7');

	redirectUrl += `fileId1=${fileId1}&fileId2=${fileId2}`;
	res.redirect(redirectUrl);
});

module.exports = router;

File: python/flask/sample/views/merge_server_files.py
import os
import uuid
from os.path import basename

from flask import Blueprint
from flask import current_app
from flask import render_template
from flask import request
from werkzeug.utils import redirect
from sample.storage_mock import get_sample_doc_name, SampleDocs, create_app_data

blueprint = Blueprint(basename(__name__), __name__,
                      url_prefix='/merge-server-files')


@blueprint.route('/<rc>', methods=['GET', 'POST'])
def index(rc):
    if request.method == 'GET':
        available_files = [
            MergeServerFileModel(0, SampleDocs.CMS_DETACHED_1, SampleDocs.CMS_DETACHED_2, 'A sample where both CMS don\'t have encapsulated content.'),
            MergeServerFileModel(1, SampleDocs.CMS_ATTACHED_1, SampleDocs.CMS_DETACHED_2, 'A sample where a CMS has encapsulated content and another doesn\'t.'),
            MergeServerFileModel(2, SampleDocs.CMS_ATTACHED_1, SampleDocs.CMS_ATTACHED_2, 'A sample where both CMS have encapsulated content.'),
            MergeServerFileModel(3, SampleDocs.CMS_DETACHED_1, None, 'A sample where the original file is attached to a CMS that doesn\'t have encapsulated content.')]
        sample_file_id = SampleDocs.CMS_DATA_FILE
        return render_template('merge_server_files/index.html', available_files=available_files, sample_file_id=sample_file_id)
    else:
        sample_id = int(request.form['selectedSample'])
        data_file = get_sample_doc_name(SampleDocs.CMS_DATA_FILE)

        redirect_url = '../../' + rc + "/"

        if sample_id == 0:
            file1 = get_sample_doc_name(SampleDocs.CMS_DETACHED_1)
            file2 = get_sample_doc_name(SampleDocs.CMS_DETACHED_2)

            redirect_url += copy_from_static_to_app_data(data_file) + \
                      '/' + copy_from_static_to_app_data(file1)     + \
                      '/' + copy_from_static_to_app_data(file2)
        elif sample_id == 1:
            file1 = get_sample_doc_name(SampleDocs.CMS_ATTACHED_1)
            file2 = get_sample_doc_name(SampleDocs.CMS_DETACHED_2)

            redirect_url += copy_from_static_to_app_data(data_file) + \
                      '/' + copy_from_static_to_app_data(file1)     + \
                      '/' + copy_from_static_to_app_data(file2)
        elif sample_id == 2:
            file1 = get_sample_doc_name(SampleDocs.CMS_ATTACHED_1)
            file2 = get_sample_doc_name(SampleDocs.CMS_ATTACHED_2)

            redirect_url += copy_from_static_to_app_data(file1) + \
                      '/' + copy_from_static_to_app_data(file2)
        else:
            file1 = get_sample_doc_name(SampleDocs.CMS_DETACHED_1)

            redirect_url += 'attach/' + copy_from_static_to_app_data(data_file) + \
                            '/' +       copy_from_static_to_app_data(file1)

        return redirect(redirect_url)


def copy_from_static_to_app_data(filename):
    # Copy file to the App_Data folder, where the upload files is stored.
    with open(os.path.join(current_app.static_folder, filename), 'rb') as f:
        sample_file = f.read()
    # Generate a unique filename.
    new_filename = '%s.%s' % (str(uuid.uuid4()), 'p7s')

    # Move the file to the "app_data" with the unique filename. Make sure
    # the "app_data" folder exists (static/util.py).
    create_app_data()
    with open(os.path.join(current_app.config['APPDATA_FOLDER'], new_filename), 'wb') as f:
        f.write(sample_file)

    return new_filename


class MergeServerFileModel:
    def __init__(self, optionId, id1, id2, description):
        self.optionId = optionId
        self.id1 = id1
        self.id2 = id2
        self.description = description